src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -692,7 +687,9 @@ JRT_LEAF(BasicType, Deoptimization::unpack_frames(JavaThread* thread, int exec_m
   // It is actually ok to allocate handles in a leaf method. It causes no safepoints,
   // but makes the entry a little slower. There is however a little dance we have to
   // do in debug mode to get around the NoHandleMark code in the JRT_LEAF macro
-  ResetNoHandleMark rnhm; // No-op in release/product versions
+  // Beware though because allocating Handles must have a HandleMark or else the
+  // Handles will be leaked.
+  ResetNoHandleMark rnhm;
   HandleMark hm(thread);
```

Comment:
Here's a simpler rewrite for L687 -> L693:

```
    // This code wants to allocate handles which is okay in a leaf method,
    // but JRT_LEAF contains a NoHandleMark so we have to work around
    // that with ResetNoHandleMark in order to create a HandleMark here:
    ResetNoHandleMark rnhm;
    HandleMark hm(thread);
```

Reply:
Unfortunately this is not entirely what I want to say.
I want to say that you *must* have a HandleMark if you have ResetNoHandleMark in a JRT_LEAF.
You need a ResetNoHandleMark to create Handles but there are no asserts to make sure you have a HandleMark.
ie:

     ResetNoHandleMark rnhm;
     Handle h(THREAD, obj);  // OK no problem

But the Handle will leak once this function returns (until the next JRT_ENTRY anyway).

    // This code wants to allocate handles which is okay in a leaf method,
    // but JRT_LEAF contains a NoHandleMark so we have to work around
    // that with ResetNoHandleMark in order to create a Handles here.
    // There must also be a HandleMark also here, to clean up any handles created here.

How about that?

Reply:
My suggestion:
```
// JRT_LEAF methods don't normally allocate handles and there is a
// NoHandleMark to enforce that. It is actually safe to use Handles
// in a JRT_LEAF method, and sometimes desirable, but to do so we
// must use ResetNoHandleMark to bypass the NoHandleMark, and
// then use a HandleMark to ensure any Handles we do create are
// cleaned up in this scope.
```
I still question why such a function declares itself a LEAF instead of just being a JRT_ENTRY but that is a different argument. :)
Cheers,
David

Reply:
Ok, your comment says the same thing as mine and includes what I want to say, so I'll use your comment.
I'm pretty sure this function wants a NoSafepointVerifier but some of the calls it uses might need handles, because we were pretty aggressive about making arguments into handles.

Reply:
I actually think the HandleMark is there because of the PreserveExceptionMark call in VerifyStack, which is the thing that Erik ran into also with stack watermarks, and prompted the original request about lifting the restrictions for JRT_LEAF functions.  I don't want there to be an unhandleized version of PreserveExceptionMark, and I don't want JRT_LEAF to allow Handles by default, so hopefully this comment is adequate documentation.